scene/gui/control.cpp

```diff
@@ -3018,6 +2984,41 @@ void Control::_notification(int p_notification) {
 			get_viewport()->_gui_remove_control(this);
 		} break;
 
+		case NOTIFICATION_ENTER_TREE: {
+			notification(NOTIFICATION_THEME_CHANGED);
+		} break;
+
+		case NOTIFICATION_PARENTED: {
+			if (data.theme.is_valid()) {
+				// Already has a theme, don't need to search parent for the theme owner.
+			} else {
+				// Check if the parent is a Control or Window, if so then use its theme owner.
+				Control *parent_c = Object::cast_to<Control>(get_parent());
+				if (parent_c) {
+					data.theme_owner = parent_c->data.theme_owner;
+					data.theme_owner_window = parent_c->data.theme_owner_window;
+				} else {
+					Window *parent_w = Object::cast_to<Window>(get_parent());
+					if (parent_w) {
+						data.theme_owner = parent_w->theme_owner;
+						data.theme_owner_window = parent_w->theme_owner_window;
+					} else {
+						data.theme_owner = nullptr;
+						data.theme_owner_window = nullptr;
+					}
+				}
+			}
+
+			_propagate_theme_changed(this, data.theme_owner, data.theme_owner_window);
```

Comment:
I'm wondering if maybe, when a control has `theme` set, `theme_owner` should still refer to the next ancestor node with a theme. This would allow optimize here with something like (pseudo-code):
```cpp
if (new_theme_owner != current_theme_owner) {
	_propagate_theme_changed(this, theme_owner);
}
```

Then we wouldn't need to propagate theme changes to all the node's children recursively if the new parent doesn't result in any theme change. Right now that code wouldn't work as if `theme_owner` was referring to this control, then we have no way of knowing what this control's previous next-up theme owner was.

Reply:
Nevermind I figured something out